src/commands/qrcode.cpp

```diff
@@ -20,177 +20,57 @@
+#include <sstream>
 #include <iostream>
 #include <bitcoin/bitcoin.hpp>
 #include <bitcoin/explorer/define.hpp>
 
-#if defined WITH_QRENCODE && defined WITH_LIBPNG
-#include "png.h"
-#include "qrencode.h"
-#endif
-
 using namespace bc;
 using namespace bc::explorer;
 using namespace bc::explorer::commands;
 
-/*
-  The following constants and write_png method are based on:
-
-  https://github.com/fukuchi/libqrencode/blob/3.4/qrenc.c
-*/
-
-#if defined WITH_QRENCODE && defined WITH_LIBPNG
-static constexpr int dpi = 72;
-static constexpr int version = 0;
-static constexpr int margin = 2;
-static constexpr QRencodeMode hint = QR_MODE_8;
-static constexpr QRecLevel level = QR_ECLEVEL_L;
-static constexpr int casesensitive = 1;
-static constexpr int INCHES_PER_METER = (100.0 / 2.54);
-static constexpr unsigned int fg_color[4] = {0, 0, 0, 255};
-static constexpr unsigned int bg_color[4] = {255, 255, 255, 255};
-
-
-static bool write_png(QRcode *qrcode, uint32_t size, std::string outfile);
-#endif
-
 console_result qrcode::invoke(std::ostream& output, std::ostream& error)
 {
-#if defined WITH_QRENCODE && defined WITH_LIBPNG
     // Bound parameters.
     const auto& address = get_payment_address_argument();
-    const auto& filename = get_filename_argument();
+    const auto& prefix = get_prefix_option();
+    const auto& image = get_image_option();
     const auto& size = get_size_option();
 
-    std::string qr_string = "bitcoin:" + address.encoded();
-
-    QRcode* qrcode = QRcode_encodeString(
-        qr_string.c_str(), version, level, hint, casesensitive);
-
-    if (qrcode && write_png(qrcode, size, filename.c_str()))
-        return console_result::okay;
-
-    error << BX_QRCODE_WRITE_ERROR << std::endl;
-    return console_result::failure;
-#else
-    error << BX_QRCODE_REQUIRES_QRENCODE_AND_LIBPNG << std::endl;
-    return console_result::failure;
-#endif
-}
-
-#if defined WITH_QRENCODE && defined WITH_LIBPNG
-static bool write_png(QRcode *qrcode, uint32_t size, std::string outfile)
-{
-    static FILE *fp; // avoid clobbering by setjmp.
-    png_structp png_ptr;
-    png_infop info_ptr;
-    png_colorp palette;
-    png_byte alpha_values[2];
-    unsigned char *row, *p, *q;
-    int x, y, xx, yy, bit;
-    int realwidth;
-
-    realwidth = (qrcode->width + margin * 2) * size;
-    row = (unsigned char *)malloc((realwidth + 7) / 8);
-    if(row == NULL) {
-        return false;
-    }
-
-    if(outfile[0] == '-' && outfile[1] == '\0') {
-        fp = stdout;
-    } else {
-        fp = fopen(outfile.c_str(), "wb");
-        if(fp == NULL) {
-            return false;
+#ifdef WITH_QRENCODE
+    std::string qr_string = prefix + address.encoded();
+    data_chunk qr_data = bc::wallet::qrencode_data(to_chunk(qr_string));
+
+    // The image option specifies we're writing output in png format
+    if (image)
+    {
+#ifdef WITH_PNG
+        const auto& filename = get_filename_argument();
+        auto ret = false;
+        if (!filename.empty())
```

Comment:
I would eliminate the filename option. It's no simpler to provide a path name on the command line than to pipe the output to a file. This would improve consistency across commands and simplify code and documentation.